File: java/tests/java8/sodium/MemoryTest4.java
package sodium;

import sodium.*;

public class MemoryTest4
{
    public static void main(String[] args)
    {
        new Thread() {
            public void run()
            {
                try {
                    while (true) {
                        System.out.println("memory "+Runtime.getRuntime().totalMemory());
                        Thread.sleep(5000);
                    }
                }
                catch (InterruptedException e) {
                    System.out.println(e.toString());
                }
            }
        }.start();

        EventSink<Integer> et = new EventSink<Integer>();
        EventSink<Integer> eChange = new EventSink<Integer>();
        Behavior<Event<Integer>> oout = eChange.map(x -> (Event<Integer>)et).hold((Event<Integer>)et);
        Event<Integer> out = Behavior.switchE(oout);
        Listener l = out.listen(tt -> {
            System.out.println(tt);
        });
        int i = 0;
        while (i < 1000000000) {
            eChange.send(i);
            i++;
        }
        l.unlisten();
    }
}

File: clojure/sodium/src/sodium/memory_test_4.clj
(ns sodium.memory-test-4
  (:require [sodium.core :refer :all])
  (:import [nz.sodium])
  (:import [nz.sodium Operational Stream StreamSink StreamLoop Tuple2 Cell CellSink Transaction Node])
  (:import [java.util Optional])
  (:gen-class))

(defn -main
  "Memory test #4"
  [& args]
  (println "Starting!")
  (.start (Thread. (fn []
                     (loop []
                       (println (str "memory " (.totalMemory (Runtime/getRuntime))))
                       (Thread/sleep 5000)
                       (recur)))))
  (let [et (StreamSink.)
        e-change (StreamSink.)
        oout (.hold (.map e-change (apply1 [x] (et))) et)
        out (Cell/switchS oout)
        l (.listen out (handler [tt] (println tt)))]
    (loop [i 0]
      (when (< i 1000000000)
        (.send et i)
        (recur (inc i))))
    (.unlisten l))
  (System/exit 0 ))  
  
                     

